File: structural/decorator/decorator.java
public class DecoratorPattern {
    public static abstract class Cat {
        protected String breed;
        protected String name;

        public Cat(String name) {
            this.name = name;
        }

        public String describe() {
            return String.format("Cat %s\nDescription: %s", name, breed);
        }
    }

    public static class Siamese extends Cat {
        public Siamese(String name) {
            super(name);
            breed = "Siamese";
        }
    }

    public static class MaineCoon extends Cat {
        public MaineCoon(String name) {
            super(name);
            breed = "Maine Coon";
        }
    }

    public static abstract class CatDecorator extends Cat {
        protected String breed = null;
        protected Cat cat;

        public CatDecorator(Cat cat) {
            super(null);
            this.cat = cat;
        }

        public abstract String describe();
    }

    public static class Sleeping extends CatDecorator {
        public Sleeping(Cat cat) {
            super(cat);
        }

        public String describe() {
            return String.format("%s, sleeping", cat.describe());
        }
    }

    public static class Purring extends CatDecorator {
        public Purring(Cat cat) {
            super(cat);
        }
        public String describe() {
            return String.format("%s, purring", cat.describe());
        }
    }

    public static void main(String[] args) {
        Cat simba = new Siamese("Simba");
        Cat oscar = new MaineCoon("Oscar");

        simba = new Purring(simba);
        simba = new Sleeping(simba);

        System.out.println(simba.describe());
        System.out.println(oscar.describe());
    }
}

File: structural/decorator/decorator.cs
using System;

class DecoratorPattern
{
    abstract class Cat
    {
        protected abstract string breed { get; }
        protected string name { get; }

        public Cat(string name)
        {
            this.name = name;
        }

        public virtual string Describe()
        {
            return String.Format("Cat {0}\nDescription: {1}", name, breed);
        }
    }

    class Siamese : Cat
    {
        protected override string breed { get; } = "Siamese";

        public Siamese(string name) : base(name) { }
    }

    class MaineCoon : Cat
    {
        protected override string breed { get; } = "Maine Coon";

        public MaineCoon(string name) : base(name) { }
    }

    abstract class CatDecorator : Cat
    {
        protected override string breed { get; } = null;
        protected Cat cat;

        public CatDecorator(Cat cat)
            : base(null)
        {
            this.cat = cat;
        }

        public abstract override string Describe();
    }

    class Sleeping : CatDecorator
    {
        public Sleeping(Cat cat) : base(cat) { }

        public override string Describe()
        {
            return String.Format("{0}, sleeping", cat.Describe());
        }
    }

    class Purring : CatDecorator
    {
        public Purring(Cat cat) : base(cat) { }
    
        public override string Describe()
        {
            return String.Format("{0}, purring", cat.Describe());
        }
    }

    static void Main(string[] args)
    {
        Cat simba = new Siamese("Simba");
        Cat oscar = new MaineCoon("Oscar");

        simba = new Purring(simba);
        simba = new Sleeping(simba);

        Console.WriteLine(simba.Describe());
        Console.WriteLine(oscar.Describe());
    }
}
